lib/ansible/plugins/connection/local.py

```diff
@@ -131,8 +131,14 @@ def exec_command(self, cmd, in_data=None, sudoable=True):
         return (p.returncode, stdout, stderr)
 
     def _ensure_abs(self, path):
+        ''' join the path with optional cwd and ensure that the type stays the same '''
         if not os.path.isabs(path) and self.cwd is not None:
-            path = os.path.normpath(os.path.join(self.cwd, path))
+            is_binary = isinstance(path, binary_type)
```

Comment:
Callers of this method ultimately expect the data to be bytes.  Just ensure that this method returns bytes.  I would start this from the beginning, and the first thing in this function should be `b_path = to_bytes(path, errors='surrogate_or_strict')`.

In addition to that, to self document this, make sure that variables that are bytes are prefixed by `b_`, such as `b_path` and `b_cwd`.  The overhead of calling `to_bytes` on something that is already bytes should be minimal.

Reply:
I had somehow omitted the to_native calls in put_file. Therefore I thought the callers are expecting text and implemented the superfluous type checking.

Now all checks have been removed and return type is always bytes.

Reply:
@sivel, Can I please get a re-review for the changes?